T13754

GeorgesStavracas opened the pull request:
https://phabricator.endlessm.com/T13754

wjt commented on gnome-image-installer/pages/install/gis-install-page.c:
```diff
@@ -71,6 +74,58 @@ G_DEFINE_QUARK(install-error, gis_install_error);
+
+  gtk_widget_show (button);
+
+  /* Abort reformatting */
```
This comment doesn't correspond to the button.

wjt commented on gnome-image-installer/pages/install/gis-install-page.c:
```diff
@@ -590,19 +665,46 @@ gis_install_page_shown (GisPage *page)
+  gtk_header_bar_set_show_close_button (GTK_HEADER_BAR (titlebar), FALSE);
+
+  for (l = children; l != NULL ; l = l->next)
+    gtk_widget_hide (l->data);
```
`update_navigation_buttons` in `gis-assistant.c` already has logic for hiding navigation buttons on the last page. Rather than duplicating it here, perhaps the page can have a vfunc for whether the buttons should be hidden – or a method to call to explicitly hide the buttons?

wjt commented on gnome-image-installer/pages/install/gis-install-page.c:
```diff
@@ -590,19 +665,46 @@ gis_install_page_shown (GisPage *page)
+  titlebar = gis_assistant_get_titlebar (gis_driver_get_assistant (page->driver));
+  children = gtk_container_get_children (GTK_CONTAINER (titlebar));
 
+  gtk_header_bar_set_show_close_button (GTK_HEADER_BAR (titlebar), FALSE);
```
Similarly, `prepare_main_window` in `gnome-initial-setup/gis-driver.c` has the logic to show the close button and make the window deletable. Maybe it'd be cleaner to put a method onto GisDriver, what do you think?

wjt commented on gnome-image-installer/pages/install/gis-install-page.c:
```diff
@@ -71,6 +74,58 @@ G_DEFINE_QUARK(install-error, gis_install_error);
+  gtk_widget_set_can_default (button, TRUE);
+  gtk_widget_show (button);
+
+  /* Lets avoid accidents */
```
Let's

wjt commented on gnome-image-installer/pages/install/gis-install-page.c:
```diff
@@ -380,6 +439,8 @@ gis_install_page_copy (GisPage *page)
 
   g_thread_yield();
 
+  priv->writing = TRUE;
```
There's a potential race here, where you click the Close button, and then the main thread happens to yield, and only then is this thread scheduled, and it runs for long enough to overwrite a little bit of the disk.

Could add a mutex and a GCancellable (say); have the `delete` callback claim the mutex, check `writing`; if it is false, then cancel the GCancellable, and only then release the mutex. Here, you'd claim the mutex, check the cancellable, set `writing` and then release the mutex.

wjt replied:
Or maybe this race is so unlikely as to not matter.

wjt commented on gnome-image-installer/pages/install/gis-install-page.c:
```diff
@@ -71,6 +74,58 @@ G_DEFINE_QUARK(install-error, gis_install_error);
+                                                 GTK_DIALOG_MODAL | GTK_DIALOG_DESTROY_WITH_PARENT,
+                                                 GTK_MESSAGE_WARNING,
+                                                 GTK_BUTTONS_NONE,
+                                                 _("Abort reformatting the disk?"));
```
I think the verb should be consistent throughout the message / secondary text / button, but at the moment it's "abort" / "cancelling" / "stop". I think "Stop" is the clearest.

GeorgesStavracas committed:
page: add class fields to control header behavior

The headerbar and the pages currently have a hard time
communicating with each other because we have to access
widgets directly. We can't fine tune, for example, whether
we want to show only the forward or the backward button.

This commit fine tunes the GisPage control over the header
by adding class fields that are read by GisAssistant.

https://phabricator.endlessm.com/T13754

wjt commented on gnome-image-installer/pages/install/gis-install-page.c:
```diff
@@ -590,7 +590,9 @@ gis_install_page_shown (GisPage *page)
-  gchar *msg = g_strdup_printf (_("Step %d of %d"), 1, 2);
+  gchar *msg;
+
+  msg = g_strdup_printf (_("Step %d of %d"), 1, 2);
```
Not a merge blocker but I don't really think this change was necessary.

wjt commented on gnome-image-installer/pages/install/gis-install-page.c:
```diff
@@ -433,7 +494,21 @@ gis_install_page_copy (GisPage *page)
+   */
+  if (priv->warning_dialog)
+    {
+      gtk_dialog_response (GTK_DIALOG (priv->warning_dialog), GTK_RESPONSE_CANCEL);
```
Are you sure it's safe to do this from a non-main thread? Shouldn't this go into `gis_install_page_teardown`?

GeorgesStavracas committed:
install: hide header widgets on Install page

When the user reaches the Install page, changing to the
previous page is not possible anymore. But we keep showing
a sensitive Previous button that breaks whatever we're
doing.

The previous commit added class-wide fields to fine tune this
behavior, and this is the oportunity for us to use them.

Fix that by setting the class fields in such a way that the
back, next, and window control buttons are hidden.

https://phabricator.endlessm.com/T13754

GeorgesStavracas committed:
installer: ask permission to quit reformatter while copying data

When copying data, a hard quit (e.g. using the Quit context menu
from the taskbar) means that the disk is left with possibly and
probably broken contents, and that almost certainly means unusable
computer.

Ideally, the reformatter/installer must ask if the user really wants
to do that, and explicitly say that the device might be in a broken
state after that.

Fix that by adding a confirmation dialog when the window tries to be
deleted.

https://phabricator.endlessm.com/T13754

wjt reviewed (approved):
Looks good!
